### ISSUE TITLE
Document publish event tracking

### ISSUE THREAD

**Issue by Green-seeker-of-interesting** [wagtail/wagtail#10283]
### Description of the problem
The system has several public collections with documents. They must be public, because. site users should be able to download the published document and registration on the site is not provided.

When a file is uploaded to the system, it becomes available for download, regardless of whether a page with a link to it is published or not. All links to documents on pages are located in the StreamField and there is no convenient way to get document instances.

### Solutions
There are two solutions.
1. Mechanism for publishing documents by analogy with publishing pages
2. The ability to get a list of documents linked to the page, for the possibility of a software ban on downloading.

### Solution implemented in my system
It was not possible to redefine the document model in my situation because the system was already running and contained many files.
I used the following solution:
- Document registration based on post_save signal
- Load blocking based on document_served signal
- Tracking file publishing based on the page_published hook

### Сontext
Project Libraries
wagtail-2.15.3
django-3.1.13

Downloading any document is available at the address with the following pattern {url}/documents/{id}/{filename}
For the test I checked on the newest version CMS, the problem is relevant.

**Comment by thibaudcolas** [comment]
Hi @Green-seeker-of-interesting, I’m having trouble understanding what you’re requesting.

> Mechanism for publishing documents by analogy with publishing pages

Just like pages, documents are already available for download on save, regardless of whether there is a link to the documents on any page. But just like pages – it’s certainly easier to find docs if they are linked from _somewhere_.

> The ability to get a list of documents linked to the page, for the possibility of a software ban on downloading.

I don’t understand what you mean by "software ban" here. Wagtail has a [reference index](https://docs.wagtail.org/en/stable/advanced_topics/reference_index.html) feature, recently released, which you might be able to use for this.

**Comment by Green-seeker-of-interesting** [comment]
Исходный текст
Hello @thibaudcolas !
Let me try again to describe the scenario. The system has moderators and editors.
1. The editor creates a page and loads a document for it
2. The page has been sent for moderation
3. The moderator checks the page and if everything is in order, the page is published.

The problem is that in the period between upload and moderation, the file can potentially contain incorrect information, the disclosure of which will harm the company.

Our goal was to block the loading of the document in the interval between its loading into the system and moderation. In essence, it is necessary to prevent anonymous users from downloading documents that are not linked from at least one published page.

Thanks for the index reference link!

**Comment by thibaudcolas** [comment]
👍 thank you for the clarification, that makes a lot of sense to me. I don’t have a good sense of whether this is something we think makes sense to do in Wagtail out of the box, or as an opt-in feature, or whether we would recommend building this as a separate package. Let’s see what others think!
